haas/ext/auth/database.py

```diff
@@ -0,0 +1,147 @@
+"""Auth plugin using usernames & passwords in the DB, with HTTP basic auth.
+
+Includes API calls for managing users.
+"""
+from haas import api, model, auth
+from haas.auth import get_auth_backend
+from haas.rest import rest_call, local
+from haas.errors import *
+from sqlalchemy import Column, ForeignKey, String, Boolean, Table
+from sqlalchemy.orm import relationship
+from passlib.hash import sha512_crypt
+from schema import Schema, Optional
+import flask
+
+
+class User(model.Model):
+    """A user of the HaaS.
+
+    A user can be a member of any number of projects, which grants them access
+    to that process's resources. A user may also be flagged as an administrator.
+    """
+
+    is_admin = Column(Boolean, nullable=False, default=False)
+    # The user's salted & hashed password. We currently use sha512 as the
+    # hashing algorithm:
+    hashed_password = Column(String)
+
+    # The projects of which the user is a member.
+    projects = relationship('Project', secondary='user_projects', backref='users')
+
+    def __init__(self, label, password, is_admin=False):
+        """Create a user `label` with the specified (plaintext) password."""
+        self.label = label
+        self.is_admin = is_admin
+        self.set_password(password)
+
+    def verify_password(self, password):
+        """Return whether `password` is the user's (plaintext) password."""
+        return sha512_crypt.verify(password, self.hashed_password)
+
+    def set_password(self, password):
+        """Set the user's password to `password` (which must be plaintext)."""
+        self.hashed_password = sha512_crypt.encrypt(password)
+
+
+# A joining table for users and projects, which have a many to many relationship:
+user_projects = Table('user_projects', model.Base.metadata,
+                      Column('user_id', ForeignKey('user.id')),
+                      Column('project_id', ForeignKey('project.id')))
+
+
+@rest_call('PUT', '/auth/basic/user/<user>', schema=Schema({
+    'password': basestring,
+    Optional('is_admin'): bool,
+}))
+def user_create(user, password, is_admin=False):
+    """Create user with given password.
+
+    If the user already exists, a DuplicateError will be raised.
+    """
+    get_auth_backend().require_admin()
+
+    # XXX: We need to do a bit of refactoring, so this is available outside of
+    # haas.api:
+    api._assert_absent(User, user)
```

Comment:
Yeah, the function is badly named; it's been that way for a *looong* time.

meta point: we really ought to dedicate some small amount of resources to this sort of thing. It doesn't belong as part of this pr, but this is trivial to fix. Given the way we've done development though it'll never happen, because it's never going to be a priority.

See also:

https://en.wikipedia.org/wiki/Paper_cut_bug

/rant

But again, this is a longstanding issue.

//cc @henn